Microservice update brunch

Denisdeparture opened the pull request:
Тут неплохо бы писать краткое описание того что сделано в ПР

Denisdeparture committed:
ChangeWebClientControllersLogic

Большинство операций вынес на веб сервер

Denisdeparture committed:
AddHtmlHelper

Denisdeparture committed:
ChangeViewModelsCommit

Убрал некоторый view заменив их контрактами

Denisdeparture committed:
ModelsPathChange

Denisdeparture committed:
MinimalChangeFromStaticFile

Denisdeparture committed:
ReplaceBackValidationOnFronValidation

Denisdeparture committed:
ChangeLogiccommit

Вынес логику на веб сервер

Denisdeparture committed:
MinimalChangeFromViews

Denisdeparture committed:
MinimalChangeFromCSProjectFile

Denisdeparture committed:
AddNewExtensionInBL

Denisdeparture committed:
AddLinkEndpointSystem

Суть в том чтобы сделать процесс добавления фронтенда (хотя бы на админку) динамичным. То есть на веб сервере существует endpoint к которому можно обратиться по средством формы на фронте, но вместо того чтобы писать форму а потом добовлять её в проект в админки можно будет добавить файл сожержащий html код формы. Система пока не до конца продумана.

Denisdeparture committed:
MinimalChangeFromBLModels

Добавлена новая модель для удобного представления возвращённых данных с Продюссера

Denisdeparture committed:
AddMessageBrokerFunctional

Добавлен клиент для работы с кафкой:
1. Метод Produce для отправки данных в партиции
2. Метод Consume для прослушивания топика на наличие новых данных. После того как данные поступят в партицию в течении указанного времени они будут добавлены в коллекцию.
3. Коллекция ObservableCollection идеально подходит для работы с месседж брокерами так как помимо хранения данных содержит событие позволяющие отслеживать изменение данных

Denisdeparture committed:
MinimalChangeInYandexCloudClient

Denisdeparture committed:
AddConfigureEndPointManager

Denisdeparture committed:
MinimalChangeJwtManger

Denisdeparture committed:
MinimalChangeInRep

Denisdeparture committed:
AddInterfacesFromWorkTokafkaClient

Класс KafkaClient реализует их так что за место него можно создать RebbitMQClient ;)

Denisdeparture committed:
MinimalChangeFromIYandexCloudClient

Denisdeparture committed:
MinimalChangeWorkWithEndpoints

Denisdeparture committed:
MinimalChangeBLFile

Denisdeparture committed:
AddContracts

Denisdeparture committed:
MinimalChangeFromDBFile

Denisdeparture committed:
AddBackgroundServiceForWorkRegistration

В этом сервисе реализован функционал консьюмера

Denisdeparture committed:
MInimalChangeFromControllers

Endpoint регистрации удалён из-за ненадобности

Denisdeparture committed:
AddDockerFileFromWebServer

Denisdeparture committed:
MinimalChangeFromWebServer

Denisdeparture committed:
AddReverseProxyServer

Добавлен Nginx

Denisdeparture committed:
MinimalChangeFromFilesInServerDirectory

Denisdeparture committed:
MinimalChangeInTest

dkobzin commented on Server/Application/Controllers/AccountController.cs:
```diff
@@ -48,55 +59,43 @@ public async Task<IActionResult> LoginPost([FromForm] LoginViewModel loginmodel)
-                
-                var actionUrlFormRedirect = Url.Action("Katalog", "Main");
-                return Redirect(actionUrlFormRedirect!);
+                return Redirect(Url.Action("Katalog", "Main")!);
```
Ошибка скорее всего - надо бы Catalog

dkobzin replied:
Ниже то же исправьте

Denisdeparture replied:
Сделал

dkobzin commented on Server/Application/Controllers/AccountController.cs:
```diff
@@ -48,55 +59,43 @@ public async Task<IActionResult> LoginPost([FromForm] LoginViewModel loginmodel)
         }
         public async Task<IActionResult> ProviderAuthenticationCallback(string returnUrl)
         {
             ExternalLoginInfo? obj = await _signInManager.GetExternalLoginInfoAsync(returnUrl);
```
Лучше переменный называть не obj а по наимнованию типа с маленькой буквы - camelCase

Denisdeparture replied:
Исправил

dkobzin commented on Server/Application/HtmlHelpers/FileHtmlTagHelper.cs:
```diff
@@ -0,0 +1,27 @@
+       public static HtmlString ShowMarkingFromThisPath(this IHtmlHelper page, string pathfromhtmlfile, IWebHostEnvironment host, ILogger<Program> logger)
+       {
+            
+            if (string.IsNullOrWhiteSpace(pathfromhtmlfile)) return new HtmlString("");
```
Используйте вместо "" - String.Empty

Denisdeparture replied:
Исправил

dkobzin commented on Server/Application/HtmlHelpers/FileHtmlTagHelper.cs:
```diff
@@ -0,0 +1,27 @@
+            try
+            {
+                marking = File.ReadAllText(path);
+                if (string.IsNullOrEmpty(marking)) throw new Exception("marking was null");
```
Общие строковые переменные лучше вынести в статический класс как константы и использовать их потом везде по имени

Denisdeparture replied:
Сделал папку ConstStorage в бизнес логике

dkobzin commented on Server/WebServer/BackgroundServices/RegistrationService.cs:
```diff
@@ -0,0 +1,75 @@
+        public readonly CancellationTokenSource _tokenSource;
+        public RegistrationService(ILogger<Program> logger, IServiceProvider serviceProvider,IMessageBrokerClient<User> client,IConfiguration configuration)
+        {
+            _messageBroker = client;
+            _tokenSource = new CancellationTokenSource();
+            var scopeUserManager = serviceProvider.CreateScope();
+            _userManager = scopeUserManager.ServiceProvider.GetRequiredService<UserManager<User>>();
+            var scopeRoleManager = serviceProvider.CreateScope();
+            _roleManager = scopeRoleManager.ServiceProvider.GetRequiredService<RoleManager<IdentityRole>>(); 
+            _unitOfWork = new UnitOfWork(_userManager, _roleManager, logger);
+            _configuration = configuration;
+            _logger = logger;
+            //_serviceProvider = serviceProvider;
+            var users = _messageBroker.Models;
+            users.CollectionChanged += Users_Register;
```
Используйте DI / IoC. Часть сервисов,  которые генерируется и обьявлены как  свойства уровне класса  , не должна свойства бытьа обьявлено в конструкторе. Используйте из в методах как локальные на переменные.

Denisdeparture replied:
Сделал, пришлось отказаться от UnitOfWorkа

dkobzin commented on Server/WebServer/BackgroundServices/RegistrationService.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+        protected override Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            var task = _unitOfWork.adminOperations.CreateRole("Test");
```
Вынесите строки в отдельный класс типа  Role. Еще лучше как  Enum с приведением к строке

Denisdeparture replied:
Сделал в виде отдельного enum

dkobzin commented on Server/WebServer/Controllers/AccountEndpointsController.cs:
```diff
@@ -0,0 +1,43 @@
+        public async Task<IActionResult> Login([FromQuery]User obj)
+        {
+           
+            var user = await _userManager.FindByEmailAsync(obj.Email!);
```
Необходимо проверять приходящие параметры на  null  с выдачей соотв. исключения. То же сделайте пожалуйста во всех actions

Denisdeparture replied:
Сделал

dkobzin commented on Server/WebServer/Controllers/AdminOperationController.cs:
```diff
@@ -0,0 +1,130 @@
+        public AdminOperationController(MainDbContext database, IYandexClient client, IConfiguration configuration, IStrategyValidation cond, ILogger<Program> logger, IHostEnvironment application)
+        {
+            _productManager = new ProductOperation(database, client, configuration);
+            _strategy = cond;
```
Отсутствует проверка входной параметров на nul с выдачей исключения. То же сделайте для всех конструкторов и методов

Denisdeparture replied:
сделал

dkobzin commented:
Да и юнит тесты надо не забывать добавлять на вашу функциональности

Denisdeparture committed:
AddAdminMicroservice

Добавил админку в виде отдельного микросервиса

Denisdeparture committed:
MinimalChangeInMainController

Добавил новую view

Denisdeparture committed:
MinimalChangeInControllers

Добавил 2 новых контроллера:
1. Отвечает за пользовательский аккаунт(user)
2. Хранит в себе 4 actonа ведущие на разные view(site)

Denisdeparture committed:
MinimalChangeInModelAndFunctionality

Вынес htmlhalperа в бизнес логику добавил новое представление

Denisdeparture committed:
MinimalChangeInViews

Добавил в _layout ссылку на аккаунт пользователя

Denisdeparture committed:
RenameViews

Denisdeparture committed:
ChangeInMainController

Добавил новое представление

Denisdeparture committed:
AddTwoNewControllers

Denisdeparture committed:
MinimalChangeInImports

Denisdeparture committed:
MinimalChangeInStaticFiles

Denisdeparture committed:
ClientCommit

Denisdeparture committed:
AddConstStorage

Вынес многие строковые значения в отдельные константы

Denisdeparture committed:
ChangeUnitOfWork

Я отказался от использования паттерна unitofwork, теперь все реализации паттерна репозиторий хранятся как отдельные зависимости

Denisdeparture committed:
AddRoleEnum

Роли в виде перечисления

Denisdeparture committed:
ChangeInLogigWithWebServer

В конструкторах теперь есть проверка на null
